nav2_costmap_2d/include/nav2_costmap_2d/costmap_filters/costmap_filter.hpp

```diff
@@ -0,0 +1,118 @@
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Alexey Merzlyakov
+ *********************************************************************/
+
+#ifndef NAV2_COSTMAP_2D__COSTMAP_FILTER_HPP_
+#define NAV2_COSTMAP_2D__COSTMAP_FILTER_HPP_
+
+#include "rclcpp/rclcpp.hpp"
+
+#include <string>
+
+#include "nav2_costmap_2d/costmap_layer.hpp"
+
+namespace nav2_costmap_2d
+{
+
+class CostmapFilter : public CostmapLayer
+{
+public:
+  CostmapFilter();
+  ~CostmapFilter();
+
+  /** Layer API **/
+  virtual void onInitialize() final;
+  virtual void updateBounds(
+    double robot_x, double robot_y, double robot_yaw,
+    double * min_x, double * min_y, double * max_x, double * max_y) final;
+  virtual void updateCosts(
+    nav2_costmap_2d::Costmap2D & master_grid,
+    int min_i, int min_j, int max_i, int max_j) final;
+
+  virtual void activate() final;
+  virtual void deactivate() final;
+  virtual void reset() final;
+
+  /** CostmapFilter API **/
+  /**
+   * @brief: Initializes map filter. Creates subscriptions to a map filter related topics
+   * @param: Name of costmap filter info topic for map filter
+   */
+  virtual void initializeFilter(
+    const std::string costmap_filter_info_topic) = 0;
+
+  /**
+   * @brief: An algorithm for how to use that map's information. Fills the Costmap2D with
+   *         calculated data and makes an action based on processed data
+   * @param: Reference to a master costmap2d
+   * @param: Low window map boundary OX
+   * @param: Low window map boundary OY
+   * @param: High window map boundary OX
+   * @param: High window map boundary OY
+   * @param: Robot position x
+   * @param: Robot position y
+   * @param: Robot orientation
+   */
+  virtual void process(
+    nav2_costmap_2d::Costmap2D & master_grid,
+    int min_i, int min_j, int max_i, int max_j,
+    double robot_x, double robot_y, double robot_yaw) = 0;
+
+  /**
+   * @brief: Resets map filter. Stops all subscriptions
+   */
+  virtual void resetFilter() = 0;
+
+private:
+  /**
+   * @brief: Rolling window indicator
+   */
+  bool is_rolling_;
+
+  /**
+   * @brief: Name of costmap filter info topic
+   */
+  std::string costmap_filter_info_topic_;
+
+  /**
+   * @brief: Latest robot position
+   */
+  double latest_robot_x, latest_robot_y, latest_robot_yaw;
```

Comment:
You could instead store this as a Pose2D and then the process() function has 6 inputs rather than 8

Reply:
Done in: https://github.com/ros-planning/navigation2/pull/1882/commits/cfeeb1a0b96560bb68fa8f8119e27a62c65e7100#diff-cffa086d2532b26fb31fc174a5b5878cL92, https://github.com/ros-planning/navigation2/pull/1882/commits/cfeeb1a0b96560bb68fa8f8119e27a62c65e7100#diff-cffa086d2532b26fb31fc174a5b5878cL113, https://github.com/ros-planning/navigation2/pull/1882/commits/cfeeb1a0b96560bb68fa8f8119e27a62c65e7100#diff-1ec3d87799ed7180cfd3235fec037076L64 and https://github.com/ros-planning/navigation2/pull/1882/commits/cfeeb1a0b96560bb68fa8f8119e27a62c65e7100#diff-f67a16397bb5979b7cbf4729ea53a451L97